python/cudf/cudf/core/udf/groupby_utils.py

```diff
@@ -0,0 +1,199 @@
+    """
+    Function to write numba kernels for `Groupby.apply` as a string.
+    Workaround until numba supports functions that use `*args`
+    """
+    # Create argument list for kernel
+    frame = _supported_cols_from_frame(
+        frame, supported_types=SUPPORTED_GROUPBY_NUMPY_TYPES
+    )
+    input_columns = ", ".join([f"input_col_{i}" for i in range(len(frame))])
+    extra_args = ", ".join([f"extra_arg_{i}" for i in range(len(args))])
+
+    # Generate the initializers for each device function argument
+    initializers = []
+    for i, colname in enumerate(frame.keys()):
+        initializers.append(
+            group_initializer_template.format(idx=i, name=colname)
+        )
+
+    return groupby_apply_kernel_template.format(
+        input_columns=input_columns,
+        extra_args=extra_args,
+        group_initializers="\n".join(initializers),
+    )
+
+
+def _get_groupby_apply_kernel(frame, func, args):
+    np_field_types = np.dtype(
+        list(
+            _supported_dtypes_from_frame(
+                frame, supported_types=SUPPORTED_GROUPBY_NUMPY_TYPES
+            ).items()
+        )
+    )
+    dataframe_group_type = _get_frame_groupby_type(
+        np_field_types, frame.index.dtype
+    )
+    return_type = _get_udf_return_type(dataframe_group_type, func, args)
+
+    # Dict of 'local' variables into which `_kernel` is defined
+    global_exec_context = {
+        "cuda": cuda,
+        "Group": Group,
+        "dataframe_group_type": dataframe_group_type,
+        "types": types,
+    }
+    kernel_string = _groupby_apply_kernel_string_from_template(frame, args)
+
+    kernel = _get_kernel(kernel_string, global_exec_context, None, func)
+
+    return kernel, return_type
+
+
+@_cudf_nvtx_annotate
+def jit_groupby_apply(offsets, grouped_values, function, *args):
+    """
+    Main entrypoint for JIT Groupby.apply via Numba.
+
+    Parameters
+    ----------
+    offsets : list
+        A list of integers denoting the indices of the group
+        boundaries in grouped_values
+    grouped_values : DataFrame
+        A DataFrame representing the source data
+        sorted by group keys
+    function: callable
+        The user UDF defined on a DataFrame
+    """
+    offsets = cp.asarray(offsets)
+    ngroups = len(offsets) - 1
+
+    kernel, return_type = _get_groupby_apply_kernel(
+        grouped_values, function, args
+    )
+    return_type = numpy_support.as_dtype(return_type)
+
+    output = cudf.core.column.column_empty(ngroups, dtype=return_type)
+    launch_args = [
+        offsets,
+        output,
+        grouped_values.index,
+    ]
+    launch_args += list(
+        _supported_cols_from_frame(
+            grouped_values, supported_types=SUPPORTED_GROUPBY_NUMPY_TYPES
+        ).values()
+    )
+    launch_args += list(args)
+
+    max_group_size = cp.diff(offsets).max()
+
+    if max_group_size >= 1000:
+        blocklim = 256
+    else:
+        blocklim = ((max_group_size + 32 - 1) // 32) * 32
```

Comment:
The 256 seems a bit odd. Won't this logic result in the blocklim ranging from 0-1000 as the max group size increases from 0, only to see a step drop from 1000 to 256 once the max group size exceeds 1000?

Reply:
@bwyogatama do you remember why we had this setup this way?

Reply:
Yeah there is no good reason honestly. I tried it with various blocklim values when the group size exceeds 1000 and I just observed that using 256 results in slightly better performance than the others.

Reply:
256 is often an ideal value for block sizes. However, the threshold of 1000 seems arbitrary. We should avoid a step function. Maybe this should be `if max_group_size >= 256`?